src/array_api_extra/_funcs.py

```diff
@@ -548,3 +559,279 @@ def sinc(x: Array, /, *, xp: ModuleType | None = None) -> Array:
     )
     return xp.sin(y) / y
+
+
+_undef = object()
+
+
+class at:  # pylint: disable=invalid-name
+    """
+    Update operations for read-only arrays.
+
+    This implements ``jax.numpy.ndarray.at`` for all backends.
+
+    Parameters
+    ----------
+    x : array
+        Input array.
+    idx : index, optional
+        You may use two alternate syntaxes::
+
+          at(x, idx).set(value)  # or get(), add(), etc.
+          at(x)[idx].set(value)
+
+    copy : bool, optional
+        True (default)
+            Ensure that the inputs are not modified.
+        False
+            Ensure that the update operation writes back to the input.
+            Raise ValueError if a copy cannot be avoided.
+        None
+            The array parameter *may* be modified in place if it is possible and
+            beneficial for performance.
+            You should not reuse it after calling this function.
+    xp : array_namespace, optional
+        The standard-compatible namespace for `x`. Default: infer
+
+    **kwargs:
+        If the backend supports an `at` method, any additional keyword
+        arguments are passed to it verbatim; e.g. this allows passing
+        ``indices_are_sorted=True`` to JAX.
+
+    Returns
+    -------
+    Updated input array.
+
+    Examples
+    --------
+    Given either of these equivalent expressions::
+
+      x = at(x)[1].add(2, copy=None)
+      x = at(x, 1).add(2, copy=None)
+
+    If x is a JAX array, they are the same as::
+
+      x = x.at[1].add(2)
+
+    If x is a read-only numpy array, they are the same as::
+
+      x = x.copy()
+      x[1] += 2
+
+    Otherwise, they are the same as::
+
+      x[1] += 2
+
+    Warning
+    -------
+    When you use copy=None, you should always immediately overwrite
+    the parameter array::
+
+        x = at(x, 0).set(2, copy=None)
+
+    The anti-pattern below must be avoided, as it will result in different behaviour
+    on read-only versus writeable arrays::
+
+        x = xp.asarray([0, 0, 0])
+        y = at(x, 0).set(2, copy=None)
+        z = at(x, 1).set(3, copy=None)
+
+    In the above example, ``x == [0, 0, 0]``, ``y == [2, 0, 0]`` and z == ``[0, 3, 0]``
+    when x is read-only, whereas ``x == y == z == [2, 3, 0]`` when x is writeable!
+
+    Warning
+    -------
+    The array API standard does not support integer array indices.
+    The behaviour of update methods when the index is an array of integers
+    is undefined; this is particularly true when the index contains multiple
+    occurrences of the same index, e.g. ``at(x, [0, 0]).set(2)``.
+
+    Note
+    ----
+    `sparse <https://sparse.pydata.org/>`_ is not supported by update methods yet.
+
+    See Also
+    --------
+    `jax.numpy.ndarray.at <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html>`_
+    """
+
+    x: Array
+    idx: Index
+    __slots__: ClassVar[tuple[str, str]] = ("idx", "x")
+
+    def __init__(self, x: Array, idx: Index = _undef, /):
```

Comment:
mypy strict mode requires this I believe

```suggestion
    def __init__(self, x: Array, idx: Index = _undef, /) -> None:
```